### PR TITLE
opencv: assorted fixes

### PR DESCRIPTION
**Pull request by aweits**
1. depends on blas when +lapack
2. set cuda nvcc flags for cuda_arch
3. let cuda/contrib builds work
4. depends on hdf5 when cuda/contrib
5. depends on ant when +java
6. allow protobuf version to be different
7. let opencv recompile it's protoc files.

### CODE REVIEW

File: var/spack/repos/builtin/packages/opencv/package.py [hunk 1]
```diff
@@ -82,6 +82,9 @@ class Opencv(CMakePackage, CudaPackage):
     variant('videoio', default=True, description='videoio module')
 
     # Optional 3rd party components
+    variant('cuda', default=False,
+            description='Activates support for CUDA, \
+                         essentially turns on +contrib')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/opencv/package.py:87]: This variant is already added by the `CudaPackage` base class, are you overriding it just to change the description?

File: var/spack/repos/builtin/packages/opencv/package.py [hunk 2]
```diff
@@ -117,8 +136,9 @@ class Opencv(CMakePackage, CudaPackage):
     depends_on('vtk', when='+vtk')
     depends_on('qt', when='+qt')
     depends_on('java', when='+java')
+    depends_on('ant', when='+java')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/opencv/package.py:139]: Is this only needed at build time?

File: var/spack/repos/builtin/packages/opencv/package.py (hunk 1 above)
Comment (aweits) [reply var/spack/repos/builtin/packages/opencv/package.py:87]: yep. all the contrib stuff comes along for the ride if you do +cuda, so felt we should let the user know - is there a better way to do that?

File: var/spack/repos/builtin/packages/opencv/package.py (hunk 2 above)
Comment (aweits) [reply var/spack/repos/builtin/packages/opencv/package.py:139]: yes, I believe so. will fix.

File: var/spack/repos/builtin/packages/opencv/package.py (hunk 1 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/opencv/package.py:87]: Nope, that's the only way. Although, is it possible to build `+cuda~contrib` or `~cuda+contrib`? If not, we should probably add a `conflicts(...)` statement.

Comment (aweits) [reply var/spack/repos/builtin/packages/opencv/package.py]: I'd really prefer to help the user out by flipping contrib on when they set cuda, but I can also see why that's (unfortunately) complicated.

File: var/spack/repos/builtin/packages/opencv/package.py (hunk 2 above)
Comment (Lyken17) [reply var/spack/repos/builtin/packages/opencv/package.py]: May I ask what if I want to turn off java support for opencv? How can specify it when `spack install`?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/opencv/package.py]: You can use:
```console
$ spack install opencv~java
```
You can also set this default in your `packages.yaml` so you don't have to specify it each time.

Comment (Lyken17) [reply var/spack/repos/builtin/packages/opencv/package.py]: thx for the information!

**Pull request merged**